x-pack/plugins/security_solution/server/usage/endpoints/index.ts

```diff
@@ -173,53 +192,71 @@ export const addEndpointDailyActivityAndPolicyDetailsToTelemetry = async (
  * to confirm whether or not the endpoint is still active
  */
 export const getEndpointTelemetryFromFleet = async (
-  savedObjectsClient: ISavedObjectsRepository
-): Promise<EndpointUsage> => {
-  // Retrieve every agent that references the endpoint as an installed package. It will not be listed if it was never installed
-  const { saved_objects: endpointAgents } = await getFleetSavedObjectsMetadata(savedObjectsClient);
+  soClient: ISavedObjectsRepository
+): Promise<EndpointUsage | {}> => {
+  // Retrieve every agent (max 10000) that references the endpoint as an installed package. It will not be listed if it was never installed
+  let endpointAgents;
+  try {
+    const response = await getFleetSavedObjectsMetadata(soClient);
+    endpointAgents = response.saved_objects;
+  } catch (error) {
+    // Better to provide an empty object rather than default telemetry as this better informs us of an error
+    return {};
+  }
+
+  const endpointAgentsCount = endpointAgents.length;
   const endpointTelemetry = getDefaultEndpointTelemetry();
 
   // If there are no installed endpoints return the default telemetry object
-  if (!endpointAgents || endpointAgents.length < 1) return endpointTelemetry;
+  if (!endpointAgents || endpointAgentsCount < 1) return endpointTelemetry;
 
   // Use unique hosts to prevent any potential duplicates
   const uniqueHostIds: Set<string> = new Set();
-  // Need agents to get events data for those that have run in last 24 hours as well as policy details
-  const agentDailyActiveTracker: AgentDailyActiveTracker = new Map();
-
-  const aDayAgo = new Date();
-  aDayAgo.setDate(aDayAgo.getDate() - 1);
   let osTracker: OSTracker = {};
+  let dailyActiveCount = 0;
+  let policyTracker: PoliciesTelemetry = { malware: { active: 0, inactive: 0, failure: 0 } };
+
+  for (let i = 0; i < endpointAgentsCount; i += 1) {
+    const { attributes: metadataAttributes } = endpointAgents[i];
+    const { last_checkin: lastCheckin, local_metadata: localMetadata } = metadataAttributes;
+    const { host, os, elastic } = localMetadata as AgentLocalMetadata; // AgentMetadata is just an empty blob, casting for our  use case
+
+    if (!uniqueHostIds.has(host.id)) {
+      uniqueHostIds.add(host.id);
+      const agentId = elastic?.agent?.id;
+      osTracker = updateEndpointOSTelemetry(os, osTracker);
+
+      if (agentId) {
+        let agentEvents;
+        try {
+          const response = await getLatestFleetEndpointEvent(soClient, agentId);
```

Comment:
Will gladly take any suggestions people may have on optimizing this, as currently there is the potential for us to make 10000 of these requests if there are that many agents.